validator/client/src/main/java/tech/pegasys/artemis/validator/client/ExternalMessageSignerService.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright 2020 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.artemis.validator.client;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import java.net.URL;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpRequest.BodyPublishers;
+import java.net.http.HttpResponse;
+import java.net.http.HttpResponse.BodyHandlers;
+import java.time.Duration;
+import org.apache.tuweni.bytes.Bytes;
+import tech.pegasys.artemis.datastructures.validator.MessageSignerService;
+import tech.pegasys.artemis.util.async.SafeFuture;
+import tech.pegasys.artemis.util.bls.BLSPublicKey;
+import tech.pegasys.artemis.util.bls.BLSSignature;
+
+public class ExternalMessageSignerService implements MessageSignerService {
+  private static final ObjectMapper MAPPER = new ObjectMapper();
+  private URL signingServiceUrl;
+  private BLSPublicKey blsPublicKey;
+  private int timeoutMs;
+
+  public ExternalMessageSignerService(
+      final URL signingServiceUrl, final BLSPublicKey blsPublicKey, final int timeoutMs) {
+    this.signingServiceUrl = signingServiceUrl;
+    this.blsPublicKey = blsPublicKey;
+    this.timeoutMs = timeoutMs;
+  }
+
+  @Override
+  public SafeFuture<BLSSignature> signBlock(final Bytes signingRoot) {
+    return sign(signingRoot, "/signer/block");
+  }
+
+  @Override
+  public SafeFuture<BLSSignature> signAttestation(final Bytes signingRoot) {
+    return sign(signingRoot, "/signer/attestation");
+  }
+
+  @Override
+  public SafeFuture<BLSSignature> signRandaoReveal(final Bytes signingRoot) {
+    return sign(signingRoot, "/signer/randao_reveal");
+  }
+
+  private SafeFuture<BLSSignature> sign(final Bytes signingRoot, final String path) {
+    try {
+      final String requestBody = createSigningRequest(signingRoot);
+      final HttpRequest request =
+          HttpRequest.newBuilder()
+              .uri(signingServiceUrl.toURI().resolve(path))
```

Comment:
It would be really useful if the signing API put the public key in the URL, not in the post data.  Load balancers can then easily split the signing load across multiple signers etc - they are dramatically more flexible with URLs than with POST data.

Reply:
This will be done in another PR, see issue https://github.com/PegaSysEng/eth2signer/issues/49